tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs

```diff
@@ -418,35 +462,58 @@ public async Task StartPlaybackAsync(string sessionId, HttpResponse outgoingResp
             DebugLogger.LogTrace($"PLAYBACK START END {id}.");
         }
 
-        public void StopPlayback(string recordingId, bool purgeMemoryStore = false)
+        public async Task StopPlayback(string recordingId, bool purgeMemoryStore = false)
         {
 
             var id = Guid.NewGuid().ToString();
             DebugLogger.LogTrace($"PLAYBACK STOP BEGIN {id}.");
 
-            if (!PlaybackSessions.TryRemove(recordingId, out var session))
+
+            // obtain the playbacksession so we can get grab a lock on it. if there is a streaming response we will HAVE TO WAIT for that to complete
+            // before we finish
+            if (!PlaybackSessions.TryGetValue(recordingId, out var session))
             {
                 throw new HttpException(HttpStatusCode.BadRequest, $"There is no active playback session under recording id {recordingId}.");
             }
 
-            if (!String.IsNullOrEmpty(session.SourceRecordingId) && purgeMemoryStore)
+            await session.Session.EntryLock.WaitAsync();
+
+            try
             {
-                if (!InMemorySessions.TryGetValue(session.SourceRecordingId, out var inMemorySession))
+                if (!PlaybackSessions.TryRemove(recordingId, out var removedSession))
                 {
-                    throw new HttpException(HttpStatusCode.InternalServerError, $"Unexpectedly failed to retrieve in-memory session {session.SourceRecordingId}.");
+                    throw new HttpException(HttpStatusCode.BadRequest, $"There is no active playback session under recording id {recordingId}.");
                 }
+                session.AuditLog.Enqueue(new AuditLogItem(recordingId, $"Lock obtained, stopping playback for {recordingId}."));
 
-                Interlocked.Add(ref Startup.RequestsRecorded, -1 * inMemorySession.Session.Entries.Count);
+                if (!AuditSessions.TryAdd(recordingId, session.AuditLog))
+                {
+                    DebugLogger.LogError($"Unable to save audit log for {recordingId}");
+                }
 
-                if (!InMemorySessions.TryRemove(session.SourceRecordingId, out _))
+                if (!String.IsNullOrEmpty(session.SourceRecordingId) && purgeMemoryStore)
                 {
-                    throw new HttpException(HttpStatusCode.InternalServerError, $"Unexpectedly failed to remove in-memory session {session.SourceRecordingId}.");
+                    if (!InMemorySessions.TryGetValue(session.SourceRecordingId, out var inMemorySession))
+                    {
+                        throw new HttpException(HttpStatusCode.InternalServerError, $"Unexpectedly failed to retrieve in-memory session {session.SourceRecordingId}.");
+                    }
+
+                    Interlocked.Add(ref Startup.RequestsRecorded, -1 * inMemorySession.Session.Entries.Count);
+
+                    if (!InMemorySessions.TryRemove(session.SourceRecordingId, out _))
```

Comment:
Couldn't just using `TryRemove(session.SourceRecordingId, out var inMemorySessions)` be sufficient here? Unless there are other locations outside of this EntryLock block which remove from the InMemorySessions which this catches a race condition between retrieval and removal.

Reply:
I think you're right. It's not like we have an early exit. This will blow up if either of them fails. May as well reduce to one check.